draft-ietf-masque-connect-udp-listen.md

```diff
@@ -154,10 +178,114 @@ UDP Payload:
 : The unmodified UDP Payload of this proxied UDP packet (referred to as "data
 octets" in {{UDP}}).
 
-# The Connect-UDP-Bind Header Field {#hdr}
+# Context ID {#contextid}
+
+This extension leverages context IDs (see {{Section 4 of CONNECT-UDP}}) to compress the target IP address and port when encoding datagrams on the wire. Either endpoint can register a context ID and the IP/ports it's associated with by sending a COMPRESSION_ASSIGN capsule to its peer. The peer will then echo that capsule to indicate it's received it. From then on, both endpoints are aware of the context ID and can send compressed datagrams. Later, any endpoint can decide to close the compression context by sending a COMPRESSION_CLOSE capsule.
+
+The context ID 0 was reserved by unextended connect-udp and is not used by this extension.
+Once an endpoint has ascertained that the peer supports this extension, the endpoint MUST NOT
+send any datagrams with context ID set to 0, and MUST drop any received datagrams with
+context ID set to 0.
+
+An even context ID MUST be used by clients to allocate uncompressed or compressed
+contexts, and an odd context ID MUST be used by the proxy to allocate compressed
+context IDs.They MAY pre-emptively use Context IDs not yet acknowledged by the other party, knowing that those packets MAY be lost since the COMPRESSION_ASSIGN request receiving proxy
+or client is not guaranteed to be ready to accept payloads until a COMPRESSION_ASSIGN
+response is echoed back.
+
+## Address Compression {#compression}
+
+The client and the proxy MAY choose to compress the IP and port information
+per datagram for a given target against the Context ID.
+In such a case, the client or the proxy MUST send a COMPRESSION_ASSIGN capsule
+(see {{capsuleassignformat}}) with the target information (see
+{{targetmappingformat}}) it wishes to compress and the other party (proxy or
+client respectively) MUST echo back with either a COMPRESSION_ASSIGN capsule
+if it accepts the compression request, or a COMPRESSION_CLOSE with the context
+ID (see {{capsulecloseformat}}) if it doesn't wish to support  compression for
+the given Context ID (For example, due to considerable memory requirements of
+establishing  a list of mappings per target per client). If the compression was
+rejected, the client and proxy MUST use an uncompressed context ID (See {{uncompressed}}) to exhange
+UDP payloads for the given target.
+
+### Uncompressed datagrams {#uncompressed}
+
+If the client wishes to allocate a Context ID for uncompressed packets,
+it MUST first exchange the COMPRESSION_ASSIGN capsule (see {{capsuleassignformat}})
+with the proxy with an unused Context ID defined in {{contextid}} with
+the IP Length set to zero.
```

Comment:
Do we still need the "Connect-UDP-Bind" header if we now require this capsule? I guess we still need it to negotiate the use of this extension, but maybe it doesn't carry a context ID any more?

Reply:
we don't. Should it be set to a boolean 0/1 value?

Reply:
Boolean SGTM. Though it's not `0`/`1` it's `?0`/`?1` Here's an example of a structured field boolean: https://www.rfc-editor.org/rfc/rfc9297#section-3.4